src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -11,45 +11,2163 @@
+							 placementUpdate->sourceNode->workerName,
+							 placementUpdate->sourceNode->workerPort,
+							 placementUpdate->targetNode->workerName,
+							 placementUpdate->targetNode->workerPort
+							 )));
+		UpdateShardPlacement(placementUpdate, responsiveWorkerList,
+							 shardReplicationModeOid);
+	}
+}
+
+
+/*
+ * SetupRebalanceMonitor initializes the dynamic shared memory required for storing the
+ * progress information of a rebalance process. The function takes a List of
+ * PlacementUpdateEvents for all shards that will be moved (including colocated
+ * ones) and the relation id of the target table. The dynamic shared memory
+ * portion consists of a RebalanceMonitorHeader and multiple
+ * PlacementUpdateEventProgress, one for each planned shard placement move. The
+ * dsm_handle of the created segment is savedin the progress of the current backend so
+ * that it can be read by external agents such as get_rebalance_progress function by
+ * calling pg_stat_get_progress_info UDF. Since currently only VACUUM commands are
+ * officially allowed as the command type, we describe ourselves as a VACUUM command and
+ * in order to distinguish a rebalancer progress from regular VACUUM progresses, we put
+ * a magic number to the first progress field as an indicator. Finally we return the
+ * dsm handle so that it can be used for updating the progress and cleaning things up.
+ */
+static void
+SetupRebalanceMonitor(List *placementUpdateList, Oid relationId)
+{
+	List *colocatedUpdateList = GetColocatedRebalanceSteps(placementUpdateList);
+	ListCell *colocatedUpdateCell = NULL;
+
+	ProgressMonitorData *monitor = CreateProgressMonitor(REBALANCE_ACTIVITY_MAGIC_NUMBER,
+														 list_length(colocatedUpdateList),
+														 sizeof(
+															 PlacementUpdateEventProgress),
+														 relationId);
+	PlacementUpdateEventProgress *rebalanceSteps = monitor->steps;
+
+	int32 eventIndex = 0;
+	foreach(colocatedUpdateCell, colocatedUpdateList)
+	{
+		PlacementUpdateEvent *colocatedUpdate = lfirst(colocatedUpdateCell);
+		PlacementUpdateEventProgress *event = rebalanceSteps + eventIndex;
+
+		strlcpy(event->sourceName, colocatedUpdate->sourceNode->workerName, 255);
+		strlcpy(event->targetName, colocatedUpdate->targetNode->workerName, 255);
+
+		event->shardId = colocatedUpdate->shardId;
+		event->sourcePort = colocatedUpdate->sourceNode->workerPort;
+		event->targetPort = colocatedUpdate->targetNode->workerPort;
+		event->shardSize = ShardLength(colocatedUpdate->shardId);
+
+		eventIndex++;
+	}
+}
+
+
+/*
+ * rebalance_table_shards rebalances the shards across the workers.
+ *
+ * SQL signature:
+ *
+ * rebalance_table_shards(
+ *     relation regclass,
+ *     threshold float4,
+ *     max_shard_moves int,
+ *     excluded_shard_list bigint[],
+ *     shard_transfer_mode citus.shard_transfer_mode,
+ *     drain_only boolean,
+ *     rebalance_strategy name
+ * ) RETURNS VOID
+ */
+Datum
+rebalance_table_shards(PG_FUNCTION_ARGS)
```

Comment:
The main challenge is that we cannot change the SQL schema, so we have to except 'auto' as a default, and 'force_logical' as an option. 

Since logical replication is not a concern we have, the behaviour of auto in community is the same as block_writes, such that parameterless rebalance_table_shards() is allowed even when there is no replica identity. Force_logical is not meaningful in community, so that gives an error that the feature is only available on Azure Database for PostgreSQL. 

Effectively community users can ignore the shard_transfer_mode parameter.